src/ffmpeg_subscriber.cpp

```diff
@@ -51,66 +83,131 @@ void FFMPEGSubscriber::subscribeImpl(
+    declareParameter(node, param_base_name, p);
+  }
+}
+
+void FFMPEGSubscriber::declareParameter(
+  rclcpp::Node * node, const std::string & base_name, const ParameterDefinition & definition)
+{
+  // transport scoped parameter (e.g. image_raw.compressed.format)
+  const auto v = definition.declare(node, base_name, getTransportName());
+  const auto & n = definition.descriptor.name;
+  if (n == "av_options") {
+    handleAVOptions(v.get<std::string>());
+  } else if (n == "measure_performance") {
+    decoder_.setMeasurePerformance(v.get<bool>());
+  } else {
+    RCLCPP_ERROR_STREAM(logger_, "unknown parameter: " << n);
+  }
+}
+
+std::string FFMPEGSubscriber::getDecodersFromMap(const std::string & encoding)
+{
+  const auto x = split_by_char(encoding, ';');
+  std::string decoders;
+  for (int i = static_cast<int>(x.size()); i > 0; --i) {
+    std::string p_name;
+    for (int j = 0; j < i; j++) {
+      p_name += "." + x[j];
+    }
+    const std::string key = param_namespace_ + "map" + p_name;
+    rcl_interfaces::msg::ParameterDescriptor pd;
+    pd.set__name(p_name)
+      .set__type(rcl_interfaces::msg::ParameterType::PARAMETER_STRING)
+      .set__description("decoders for encoding: " + p_name)
+      .set__read_only(false);
     try {
-      rcl_interfaces::msg::ParameterDescriptor pd;
-      pd.set__name(kv.first)
-        .set__type(rcl_interfaces::msg::ParameterType::PARAMETER_STRING)
-        .set__description("ffmpeg decoder map entry")
-        .set__read_only(false);
-      v = node->declare_parameter(key, rclcpp::ParameterValue(kv.second), pd);
+      decoders = node_->declare_parameter(key, rclcpp::ParameterValue(""), pd).get<std::string>();
     } catch (const rclcpp::exceptions::ParameterAlreadyDeclaredException &) {
-      RCLCPP_DEBUG_STREAM(logger_, "was previously declared: " << kv.first);
-      v = node->get_parameter(key).get_parameter_value();
+      RCLCPP_DEBUG_STREAM(logger_, "was previously declared: " << p_name);
+      decoders = node_->get_parameter_or<std::string>(key, "");
     }
-    if (v.get<std::string>() != kv.second) {
-      RCLCPP_INFO_STREAM(
-        logger_, "overriding default decoder " << kv.second << " for " << kv.first << " with "
-                                               << v.get<std::string>());
+    if (!decoders.empty()) {
+      break;
     }
   }
-  const bool mp = ffmpeg_encoder_decoder::get_safe_param<bool>(
-    node_, param_namespace_ + "measure_performance", false);
-  decoder_.setMeasurePerformance(mp);
+  return (decoders);
 }
 
-void FFMPEGSubscriber::internalCallback(const FFMPEGPacketConstPtr & msg, const Callback & user_cb)
+void FFMPEGSubscriber::handleAVOptions(const std::string & opt)
 {
-  if (!decoder_.isInitialized()) {
-    if (msg->flags == 0) {
-      return;  // wait for key frame!
-    }
-    if (msg->encoding.empty()) {
-      RCLCPP_ERROR_STREAM(logger_, "no encoding provided!");
-      return;
+  const auto split = split_by_char(opt, ',');
```

Comment:
@berndpfrommer I think I've seen this in two places. May be worth a `parseAVOptions` function